enhance `global_defs`

alexlamsl opened the pull request:
This is an alternative approach to #1467, incorporating the feature to substitute `AST_Node` when supplied via command-line (API also supported, but a bit messier).

This PR also has the advantage of not relying on #1425 thus not requiring `unsafe` to function for expanded property syntax, i.e. `process.engine="V8"` now works with default compress options.

Included (slightly modified) all the tests from #1467, added cli mocha tests for https://github.com/mishoo/UglifyJS2/pull/1467#issuecomment-277811552

/cc @kzc

kzc commented on test/compress/global_defs.js:
```diff
@@ -0,0 +1,137 @@
+        console.log(++CONFIG.VALUE);
+        console.log(++CONFIG["VAL" + "UE"]);
+        console.log(++DEBUG[CONFIG.VALUE]);
+        CONFIG.VALUE.FOO = "bar";
```
This leads to a nonsensical replacement `42..FOO` - could you replace it with something that works - be it another variable or an object literal or something?

alexlamsl replied:
Fixed - now behaves the same way as #1467.

kzc replied:
Works for me.

kzc commented:
If this PR incorporates the same tests as PR #1467 and addresses the same issues and does not rely on `unsafe`, then this PR would be preferable.

kzc commented:
I noticed that this works:
```
$ echo 'console.log(baz);' | bin/uglifyjs --define console.log=println -c
println(baz);
```
but the reverse does not:
```
$ echo 'println(baz);' | bin/uglifyjs --define println=console.log -c
undefined:8001
                && (!self.scope.uses_with || !compressor.find_parent(AST_With))) {
                               ^

TypeError: Cannot read property 'uses_with' of undefined
```
It's it's too difficult to fix we can drop the arbitrary node replacement feature altogether and just stick to replacing constants and literals.

kzc commented:
@alexlamsl By the way, these PRs are really impressive. Just tell me when you think I'm asking for an unreasonable feature.

alexlamsl commented:
@kzc as long as you are willing to test and point out areas of bugs / improvements, I'm happy to give it a go 😉 

While making fixes to #1467, I noticed there is a fundamental difference between `C={D:1}` and `C.D=1`, i.e. the former express an absolute knowledge on `C` whereas the latter is only sure `C` has one defined property with a specific value.

So this is a reboot of the efforts, separating #1425 (best-effort basis) with `global_defs` (absolute substitution). Of course in some cases both features would be required to give a final optimal result.

Let me try and address the two issues you've mentioned above, and if successful, I'll close out #1467.

kzc commented on test/compress/issue-208.js:
```diff
@@ -27,3 +27,28 @@ do_update_rhs: {
+    options = {
+        global_defs: { DEBUG: 0 }
+    }
+    input: {
```
Please add this to the beginning of input:
```
const DEBUG = true;
```
It should also generate a warning.

Could you also add `OTHER: 1` to  global_defs and the following line to the test:
```
var OTHER;
```
To ensure it also produces a warning.

alexlamsl replied:
It would generate a warning, plus killing all the positive cases below (because they are no longer `AST_SymolRef.undeclared()` 😅

alexlamsl replied:
I'll pick two `OTHER` variables as you suggested then.

kzc replied:
You could name it something else. Just curious what happens to variable declarations that are set and not set.

kzc replied:
Having `const DEBUG = true;` at the top of the file and overriding it with `--define DEBUG=false` would be a common use case. It would work well with `-c toplevel`.

alexlamsl replied:
I tried the following:
```js
global_defs: {
    DEBUG: 0,
    ENV: 1,
    FOO: 2,
}


const ENV = 3;
var FOO = 4;
f(ENV);
--FOO;
DEBUG = 1;
DEBUG++;
DEBUG += 1;
f(DEBUG);
x = DEBUG;
```
No new warnings. Do you wish to have warnings there? The example I quote in https://github.com/mishoo/UglifyJS2/pull/1467#issuecomment-277790228 would then give a warning, for instance.

alexlamsl replied:
> Having `const DEBUG = false;` at the top of the file and overriding it with `--define DEBUG=true` (or vice versa) would be a common use case.

Ah, I see. So we do want to actually override the definition? Currently (master & this PR) that `const` definition would simply make `global_defs` an no-op.

kzc replied:
I never thought about overriding the definition, although I suppose it could be overridden. Just a warning would suffice I think. It's up to you.

alexlamsl replied:
Okay, I'll go with the warning option then. Let me finish off the release and I'll get right onto it. 😉

kzc replied:
Overriding the definition would be ideal if you don't mind the extra work.

kzc replied:
> Okay, I'll go with the warning option then.

That's fine too.

alexlamsl replied:
> Overriding the definition would be ideal

Should we be limiting this override to top-level-defined variables or at any scopes?

alexlamsl replied:
So this is an interesting case to consider:
```js
global_defs: { "config.DEBUG" : 1 }

var config = { engine: "voom!" };
if (config.DEBUG) {
  log();
}
```
Should we override the value and/or raise a warning, or something else altogether?

alexlamsl replied:
I've added those tests to this PR to reflect the current behaviour. Please leave comments on what you expect each case to be and we'll go from there?

kzc commented on test/compress/issue-208.js:
```diff
@@ -30,25 +30,37 @@ do_update_rhs: {
         x = DEBUG;
     }
     expect: {
+        const ENV = 3;
```
overriding `const ENV` should warn.

alexlamsl replied:
See if https://github.com/mishoo/UglifyJS2/pull/1469/commits/d7e9eb96b93f57f324c082fc20cba881e22a7e72 works?

kzc commented on test/compress/issue-208.js:
```diff
@@ -30,25 +30,37 @@ do_update_rhs: {
     }
     expect: {
+        const ENV = 3;
+        var FOO = 4;
```
overriding `var FOO` should warn. 

One could argue in this specific case if it is set to the same value (4) it shouldn't - but that's too much work.

alexlamsl replied:
See if https://github.com/mishoo/UglifyJS2/pull/1469/commits/d7e9eb96b93f57f324c082fc20cba881e22a7e72 works?

kzc commented on test/compress/global_defs.js:
```diff
@@ -123,6 +126,8 @@ mixed: {
     }
     expect: {
+        const FOO = { BAR: 0 };
+        console.log(FOO.BAR);
```
This is a tricky case. But since `FOO` is known to global I think `FOO.BAR` should be changed to `0`.

Had `FOO` been a parameter, then it would have been left as is.

alexlamsl replied:
So do you mean no changes to current behaviour is required for this case?

kzc replied:
If `FOO` is global - which it is in this test case - then `FOO.BAR` should be changed to `0`.

alexlamsl replied:
Ah I see - the `FOO.BAR` ➡️ `0` optimisation goes under #1425. I'll add `unsafe` here to enable it then.

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/28c4d97889e8aa59c35e631c6015f08790afa344

kzc replied:
Interesting side effect. That wasn't my intention. Even if `FOO` was declared as `const FOO = null;` I was thinking that since `FOO` is global then `FOO.BAR` is fair game to be replaced with the `global_defs` setting.

alexlamsl replied:
I'm slightly confused now - `FOO.BAR` is defined as `"moo"` In `global_defs`, so do you mean you want that instead of `0`?

alexlamsl replied:
PTAL at https://github.com/mishoo/UglifyJS2/pull/1469/commits/e88acb47a941c61321a1b3710f15933f269915b2 and see if that matches what you mean?

kzc replied:
I think the new test and behavior is correct.

kzc commented on lib/compress.js:
```diff
@@ -981,6 +970,67 @@ merge(Compressor.prototype, {
     }
 
+    (function (def){
+        AST_Node.DEFMETHOD("resolve_defines", function(compressor) {
```
To speed this up when `global_defs` are not defined:
```js
        if (!compressor.option("global_defs")) return null;
```

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/00d93f3e9c6bc77081898dc53f80d13ca5445aab

kzc commented on lib/compress.js:
```diff
@@ -981,6 +970,68 @@ merge(Compressor.prototype, {
+        });
+        def(AST_SymbolRef, function(compressor, suffix){
+            if (!this.undeclared()) return;
+            var name = this.name + suffix;
```
var name;

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/85c3c6005a289bb1ec0e66a256287bf56be964ac

kzc commented on lib/compress.js:
```diff
@@ -981,6 +970,68 @@ merge(Compressor.prototype, {
+            if (!this.undeclared()) return;
+            var name = this.name + suffix;
+            var defines = compressor.option("global_defs");
+            if (defines && HOP(defines, name)) {
```
+            if (defines && HOP(defines, (name = this.name + suffix))) {

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/85c3c6005a289bb1ec0e66a256287bf56be964ac

alexlamsl commented:
@kzc you may have discovered a bug / missing feature in https://github.com/mishoo/UglifyJS2/pull/1469#discussion_r99982699
```
>echo const a={b:0};console.log(a.b) | bin/uglifyjs -c unsafe
const a={b:0};console.log(0);

>echo const a=0;console.log(a) | bin/uglifyjs -c
const a=0;console.log(a);

>echo const a=0;console.log(a) | bin/uglifyjs -c unsafe
const a=0;console.log(a);
```

kzc commented on test/compress/global_defs.js:
```diff
@@ -0,0 +1,143 @@
+    }
+    input: {
+        function f(CONFIG) {
+            return CONFIG.VALUE;
```
add comment: `// CONFIG not global - do not replace`

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/889e8798914f025cda388a21cbe833ea4d5102c2

kzc commented on test/compress/global_defs.js:
```diff
@@ -0,0 +1,143 @@
+        }
+        function g() {
+            var CONFIG = { VALUE: 1 };
+            return CONFIG.VALUE;
```
add comment: `// CONFIG not global - do not replace`

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/889e8798914f025cda388a21cbe833ea4d5102c2

kzc commented on test/compress/global_defs.js:
```diff
@@ -0,0 +1,143 @@
+    }
+    input: {
+        function f(CONFIG) {
+            return CONFIG.VALUE;
```
add comment: `// CONFIG not global - do not replace`

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/889e8798914f025cda388a21cbe833ea4d5102c2

kzc commented on test/compress/global_defs.js:
```diff
@@ -0,0 +1,143 @@
+        }
+        function g() {
+            var CONFIG = { VALUE: 1 };
+            return CONFIG.VALUE;
```
add comment: `// CONFIG not global - do not replace`

alexlamsl replied:
Done https://github.com/mishoo/UglifyJS2/pull/1469/commits/889e8798914f025cda388a21cbe833ea4d5102c2

alexlamsl commented:
@kzc I've updated `test/compress/issue-208.js` in https://github.com/mishoo/UglifyJS2/pull/1469/commits/749587501fb06eb484476f57a3754444fa7d9eb8 to clarify the value of `ENV` - is that what you expect it to be?

alexlamsl commented:
(Travis CI timed out.)

kzc commented:
I think we have keep backwards compatibility with `-c --define`:
```
$ uglifyjs -V
uglify-js 2.7.5
```
```
$ echo 'f(ENV * 10);' | uglifyjs -c --define ENV=1
f(10);
```
```
$ echo 'const ENV = 3; f(ENV * 10);' | uglifyjs -c --define ENV=1
const ENV=3;f(30);
```
But this seems inconsistent to me. ENV is global in both cases, but only in the former case is it replaced.

So if we are to keep backwards compatibility then we can't have a global `const DEBUG = false;` overridden by `-c --define`. That's unfortunate.

What's the logic governing the current behavior in master?

alexlamsl commented:
Both `2.7.5` and `master` are governed by [this rule](https://github.com/mishoo/UglifyJS2/blob/1eaa211e0932105439d98d4f03a981f157f0a77c/lib/compress.js#L2689), which means a variable will never trigger `global_defs` if defined (local or global).

So this PR current is backward compatible with `2.7.5`, with the addition of a new warning emitted on this case. I suppose if we ever go to `3.x` then we can change this behaviour?

alexlamsl commented:
FYI, `global_defs` is introduced in https://github.com/mishoo/UglifyJS2/commit/dde5b22b5e99d4a75c8918659b7f73abed436ce2, and the only change to that line since then was #208.

kzc commented:
> So this PR current is backward compatible with 2.7.5, with the addition of a new warning emitted on this case. I suppose if we ever go to 3.x then we can change this behaviour?

It's probably prudent to retain this behavior in 2.x and revisit the logic in 3.x.

If the variable is declared in 2.x, then no substitutions can take place at all - correct? Hmm. I wonder if any code in the wild would break since using the `--define` would have no effect, and as such would probably not be used.

alexlamsl commented:
> If the variable is declared in 2.x, then no substitutions can take place at all - correct?

Except for the expanded syntax, e.g. `--define CONFIG.DEBUG=false` - in this case any `CONFIG.DEBUG` will get overridden even if a global variable `CONFIG` is defined.

>Hmm. I wonder if any code in the wild would break

I'm wondering about that too - the `2.7.5` behaviour is so limited it's either not used, or the results aren't optimal. But hey, may be we should bet on the safe side, release the current form into the wild, then wait and see if there are request for enhancements on the Issue Tracker?

kzc commented:
I tend to agree. A case could easily be made to allow `--define` overrides for declared globals. We don't have to wait for another user to request it since it's clearly a useful scenario.

@avdg, @rvanvelzen Do you have an opinion on this?

alexlamsl commented:
Squashed all the previous commits, and added one to demonstrate the potential breaking change we've discussed.

FWIW, I'd vote towards including the change as it seems to be the expected behaviour, but I'm fine with it either way.

kzc commented:
LGTM

avdg commented:
LGTM unless there are coverage holes (which is a hard question anyway now especially without cover analysis).

But there are many tests in already :-)
